pulp_docker/docker_convert.py

```diff
@@ -0,0 +1,256 @@
+    signatures = [dict(header=header, signature=signature, protected=protectedHeader)]
+    jsig = _jsonDumps(dict(signatures=signatures))[1:-2]
+    arr = [jdata[:-2], ',', jsig, jdata[-2:]]
+    # Add the signature block at the end of the json string, keeping the
+    # formatting
+    jdata2 = ''.join(arr)
+    return jdata2
+
+
+def validate_signature(signed_mf):
+    """
+    Validate the signature of a signed manifest
+    """
+    # A signed manifest is a JSON document with a signature attribute
+    # as the last element.
+    # In order to validate the signature, we need the exact original payload
+    # (the document without the signature). We cannot json.load the document
+    # and get rid of the signature, the payload would likely end up
+    # differently because of differences in field ordering and indentation.
+    # So we need to strip the signature using plain string manipulation, and
+    # add back a trailing }
+
+    # strip the signature block
+    payload, sep, signatures = signed_mf.partition('   "signatures"')
+    # get rid of the trailing ,\n, and add \n}
+    jw_payload = payload[:-2] + '\n}'
+    # base64-encode and remove any trailing =
+    jw_payload = base64.urlsafe_b64encode(jw_payload.encode('ascii')).decode('ascii').rstrip("=")
+    # add payload as a json attribute, and then add the signatures back
+    complete_msg = payload + '   "payload": "{}",\n'.format(jw_payload) + sep + signatures
+    _jws = jws.JWS()
+    _jws.verify_json(complete_msg.encode('ascii'))
+
+
+def getKeyId(key):
+    """
+    DER-encode the key and represent it in the format XXXX:YYYY:...
+    """
+    derRepr = toDer(key)
+    shaRepr = hashlib.sha256(derRepr).digest()[:30]
+    b32Repr = base64.b32encode(shaRepr).decode()
+    return ':'.join(byN(b32Repr, 4))
+
+
+def toDer(key):
+    """Return the DER-encoded representation of the key"""
+    point = b"\x00\x04" + number2string(key.x, key.curve.bytes) + \
+        number2string(key.y, key.curve.bytes)
+    der = ecdsa.der
+    curveEncodedOid = der.encode_oid(1, 2, 840, 10045, 3, 1, 7)
+    return der.encode_sequence(
+        der.encode_sequence(ecdsa.keys.encoded_oid_ecPublicKey, curveEncodedOid),
+        der.encode_bitstring(point))
+
+
+def byN(strobj, N):
+    """
+    Yield consecutive substrings of length N from string strobj
+    """
+    it = iter(strobj)
+    while True:
+        substr = ''.join(itertools.islice(it, N))
+        if not substr:
+            return
+        yield substr
+
+
+def number2string(num, order):
+    """
+    Hex-encode the number and return a zero-padded (to the left) to a total
+    length of 2*order
+    """
+    # convert to hex
+    nhex = "%x" % num
+    # Zero-pad to the left so the length of the resulting unhexified string is order
+    nhex = nhex.rjust(2 * order, '0')
+    return binascii.unhexlify(nhex)
+
+
+if __name__ == '__main__':
+    sys.exit(main())
```

Comment:
Instead of calling `main()` from the script itself, the invocation should be initiated by a registry (https://github.com/pulp/pulp_docker/blob/56b12fa72906d56e5760acd982082fc697a345a1/pulp_docker/app/registry.py#L155).

Reply:
Or you can delete the function `main()` too. I think that it was used mainly for debugging purposes and it is not necessary in production. Correct me if I was wrong.

Reply:
You are correct. You are free to remove it after the patch is merged. I think having a simple way to test the conversion from the command line is valuable, and while the module itself is not the right place to put it, I don't know where to put it within the code base.